resources/web/wwi/mouse_events.js

```diff
@@ -0,0 +1,368 @@
+      case THREE.MOUSE.MIDDLE:
+        this.state.mouseDown |= 4;
+        break;
+      case THREE.MOUSE.RIGHT:
+        this.state.mouseDown |= 2;
+        break;
+    }
+
+    if (this.state.mouseDown !== 0) {
+      var relativePosition = MouseEvents.convertMouseEventPositionToRelativePosition(this.sceneManager.renderer, event.clientX, event.clientY);
+      var screenPosition = MouseEvents.convertMouseEventPositionToScreenPosition(event.clientX, event.clientY);
+      this.intersection = this.sceneManager.pick(relativePosition, screenPosition);
+
+      document.addEventListener('mousemove', this.onmousemove, false);
+      document.addEventListener('mouseup', this.onmouseup, false);
+    }
+
+    if (typeof webots.currentView.onmousedown === 'function')
+      webots.currentView.onmousedown(event);
+  },
+
+  _onMouseMove: function(event) {
+    if (!this.enableNavigation && event.button === 0) {
+      if (typeof webots.currentView.onmousemove === 'function')
+        webots.currentView.onmousemove(event);
+      return;
+    }
+
+    if (typeof this.state.x === 'undefined')
+      // mousedown event has not been called yet.
+      // This could happen for example when another application has focus while loading the scene.
+      return;
+    if ('buttons' in event)
+      this.state.mouseDown = event.buttons;
+    else if ('which' in event) { // Safari only
+      switch (event.which) {
+        case 0: this.state.mouseDown = 0; break;
+        case 1: this.state.mouseDown = 1; break;
+        case 2: this.state.pressedButton = 4; break;
+        case 3: this.state.pressedButton = 2; break;
+        default: this.state.pressedButton = 0; break;
+      }
+    }
+    if (this.state.mouseDown === 0)
+      return;
+
+    if (this.state.initialTimeStamp === null)
+      // Prevent applying mouse move action before drag initialization in mousedrag event.
+      return;
+
+    var params = {};
+    params.dx = event.clientX - this.state.x;
+    params.dy = event.clientY - this.state.y;
+
+    if (this.state.mouseDown === 1) { // left mouse button to rotate viewpoint
+      if (this.intersection && this.intersection.object) {
+        params.pickPosition = new THREE.Vector3();
+        this.intersection.object.getWorldPosition(params.pickPosition);
+      } else
+        params.pickPosition = null;
+      this.sceneManager.viewpoint.rotate(params);
+    } else {
+      if (this.intersection == null) {
+        var cameraPosition = new THREE.Vector3();
+        this.sceneManager.viewpoint.camera.getWorldPosition(cameraPosition);
+        params.distanceToPickPosition = cameraPosition.length();
+      } else
+        params.distanceToPickPosition = this.intersection.distance;
+      if (params.distanceToPickPosition < 0.001) // 1 mm
+        params.distanceToPickPosition = 0.001;
+      // FIXME this is different from webots. We need to understand why the same formula doesn't work.
+      // THREEjs Camera fov is half of Webots Viewpoint fov.
```

Comment:
I believe it is important to understand why we don't use the same formula here, as this may hide a more fundamental problem.

Reply:
Yes, but note that this was exactly the same with x3dom: we were not using the same formula as in Webots.

Reply:
The only difference I found w.r.t. Webots was that in webots.js the partial translation was applied based on the relative mouse move distance (difference from previous position) but in Webots the distance to the initial mouse position when the drag was started.
I changed the behavior in webots.js and in the world I tested I get the same behavior.